### ISSUE TITLE
Cluster member suddenly starts eating up all memory and becomes unresponsive until restarted or killed by OOM

### ISSUE THREAD

**Issue by mt-caret** [lxc/lxd#7630]
# Required information

 * Distribution: Ubuntu
 * Distribution version: 18.04
 * The output of "lxc info": https://gist.github.com/mt-caret/6f759b678be03b160eacd84ce0b9dabb

# Issue description

We run a LXD cluster consisting of 11 nodes, and we have recently been seeing a number of nodes suddenly becoming offline (as seen in `lxc cluster ls`) and the machines in question eating up an unbounded amount of RAM until swap is completely used up and sshing in becomes impossible. Restarting the lxd daemon if accessible, or restarting the server itself seems to fix the problem.

journalctl seems to show LXD stuck in some kind of deadlock situation:
https://drive.google.com/file/d/1_l4Git1JcVgnnPob799s3pRv_vy8BT24/view?usp=sharing

# Steps to reproduce

We have seen this issue two times, and but the sheer amount of logs generated causes all logs to be truncated, so we have no idea how exactly this happens. The first time it happened seemed to have been triggered by restarting the LXD daemon on another node (to try and fix some another LXD issue where the node had 50,000 simultaneous connections open to other nodes on the cluster, slowing down networking for containers to a crawl). In what I believe was a matter of minutes, the database nodes ate up 64GB of RAM and became unresponsive. Here's a screenshot of htop running on a node which I was fortunate enough to be able to ssh into.

![2020-06-21_12:51:11](https://user-images.githubusercontent.com/4996739/86742040-c90b7f00-c072-11ea-8ad1-4c9154274ab1.png)

We have no idea as to why this happened again (today, at around 5AM JST judging from our metrics), as we weren't doing maintenance/administration on the cluster at that time. dmesg output (linked below) shows OOM kicking in maybe 6 hours later, triggering a restart of LXD, "fixing" the issue.

# Information to attach

 - [x] Any relevant kernel output (`dmesg`): https://gist.github.com/mt-caret/ba03cae32fef94afa752d2d5bfed255e
 - [ ] Container log (`lxc info NAME --show-log`)
 - [ ] Container configuration (`lxc config show NAME --expanded`)
 - [x] Main daemon log (at /var/log/lxd/lxd.log or /var/snap/lxd/common/lxd/logs/lxd.log): https://gist.github.com/mt-caret/f85fb6a19c3a0dc4678926f0d576e5af
 - [ ] Output of the client with --debug
 - [ ] Output of the daemon with --debug (alternatively output of `lxc monitor` while reproducing the issue)

**Comment by freeekanayaka** [comment]
The tracebacks in the journalctl output seem to point to some kind of busy loop happening in `events/events.go`. @stgraber can you think of anything we changed recently in that area? Or it might be just a side effect of something else.

**Comment by stgraber** [comment]
We haven't done anything new there in a while that I can think of. Last change in that package was from February.

Looks like some kind of event loop or something maybe? Could be useful to run `lxc monitor --type=logging --pretty` on an affected node to see if there's a flood of event going on there?

**Comment by mt-caret** [comment]
Unfortunately, this happens sporadically which makes it difficult to predict in advance which node would be affected (we detect it after it happens, but usually it's too late and we can't ssh in). If we manage to get access to a node while it's happening, I'll try that command.

**Comment by mt-caret** [comment]
This happened again a few minutes ago, and we were able to run `lxc monitor --type=logging --pretty` on all nodes (incl. the effected ones) along with some other interesting logs. We're still in the process of dealing with the situation, so I'll post more information afterwards. Note that we have since upgraded to LXD 4.4 and this problem continues to occur.

**Comment by stgraber** [comment]
@mt-caret any luck?

**Comment by mt-caret** [comment]
@stgraber Apologies for the late response. I've recently changed jobs and have been relatively busy with relocating. 

I have the lxc monitor logs and the netstat output showing a large number of connections on effected nodes. Given the sensitive nature of the logs I think it's best if I contact you directly, is there some preferred email address that I can send the logs to?

**Comment by stgraber** [comment]
stgraber at ubuntu dot com

**Comment by mt-caret** [comment]
I've sent the logs.

**Comment by stgraber** [comment]
I've made a bit of progress on this trying to track down goroutines and potentially associated memory and network connections.
I don't have a system with such a bad case of it but I'm looking at what number of goroutines I have here.

**Comment by stgraber** [comment]
I've sent fixes for a few issues I've tracked down. There seems to be a much larger issue around dqlite leaking goroutines and likely keeping connections active. I provided the details to that to @freeekanayaka who can hopefully shed some light on it.

@mt-caret if you can identify a server with what looks like more connections than normal, can you do:
 - lxc config set core.debug_address 127.0.0.1:8444
 - curl "http:/127.0.0.1:8444/debug/pprof/goroutine?debug=2" > goroutines

And provided us that `goroutines` file? It should not contain any sensitive data, just the stack trace of all running goroutines on your system. This should make it very easy for us to see what's running within LXD which may be keeping things open.

**Comment by mt-caret** [comment]
I've since moved jobs and countries; adding to the discussion the people currently responsible for the lxd cluster @190ikp @nkpoid

**Comment by stgraber** [comment]
Thanks!

Sadly there's really not much for us to go on with here. I did a complete audit of all our goroutines on a few of our larger clusters and fixed a few un-needed ones, but those wouldn't really cause growth overtime.

I'm going to close this issue for now as there's nothing actionable for us at this time.

Should it happen again, I'd recommend following the steps I just posted in another issue: https://github.com/lxc/lxd/issues/6362#issuecomment-730024517

This on top of the usual `ps fauxww` and `netstat -taunp` should give us a good view of what's happening in a misbehaving LXD instance, hopefully letting us figure out whatever issue is happening.
